### PR TITLE
Simple client/server atari.

### PR DESCRIPTION
**Pull request by fchirica**

### CODE REVIEW

**Comment by almogdepaz** [comment]
we are missing the part where we validate the nodes against the onchain tree, that should be by using a method in data_layer.py, rn you have the get_pairs method there (i think that gives you only the terminal nodes) but maybe we should discuss whats the best way to validate

File: chia/data_layer/data_layer_client.py
Comment (almogdepaz) [inline chia/data_layer/data_layer_client.py:49]: we don't really need to send the row hash from the server we can calculate it here.
validation we should do is against the hashes onchain, we need to use the wallet through the data_layer service to to fetch those and than compare and make sure they match

File: chia/data_layer/data_store.py
Comment (arvidn) [inline chia/data_layer/data_store.py:534]: it seems to me like there should be a sql index for the key, and we should just ensure it's unique. That seems a lot cheaper than `get_pairs()`. If we have that, do we still need a way to skip the key uniqueness check?

File: chia/data_layer/data_store.py
Comment (arvidn) [inline chia/data_layer/data_store.py:404]: why do we need to lock the database for a read operation? All nodes in our database are immutable, right? There's no chance of data races or nodes disappearing, right?

File: chia/data_layer/data_store.py
Comment (arvidn) [inline chia/data_layer/data_store.py:738]: I have a feeling there's a better name for this function. What does it do exactly?
Does it return the subtree rooted in `node_hash` in the tree `tree_id` (is that the same as launcher ID?)
But why does it need the `root_hash`? The subtree is *defined* by the `node_hash`, it doesn't matter what the root is. But maybe I misunderstand what this function does.

File: chia/data_layer/data_layer_client.py
Comment (arvidn) [inline chia/data_layer/data_layer_client.py:31]: did you mean to look up the "root_hash" key here?

File: chia/data_layer/data_layer_client.py
Comment (arvidn) [inline chia/data_layer/data_layer_client.py:26]: what's the purpose of this `tree_id` constant?

File: chia/data_layer/data_layer_client.py
Comment (arvidn) [inline chia/data_layer/data_layer_client.py:14]: what is the purpose of this class? is it a test? It kind of looks like a test except it's not picked up by `pytest`.

File: chia/data_layer/data_layer_client.py
Comment (arvidn) [inline chia/data_layer/data_layer_client.py:85]: this doesn't look right. it looks like you're "flattening" the tree and compute the tree hash of a node of hashes, not a node of subtrees. isn't there a function to compute the tree hash of the data store that you can use here instead of sticking this algorithm in here directly? Either way, I think it should be factored out into a function so you can unit test it.

File: chia/data_layer/data_layer_server.py
Comment (arvidn) [inline chia/data_layer/data_layer_server.py:39]: is this supposed to be production code?

File: chia/data_layer/data_layer_server.py
Comment (arvidn) [inline chia/data_layer/data_layer_server.py:16]: is "generation" a counter we increment by 1 every time we update the singleton?

File: chia/data_layer/data_store.py
Comment (arvidn) [inline chia/data_layer/data_store.py:86]: I don't thinks we should store hashes as text, use blob instead

File: tests/core/data_layer/util.py
Comment (arvidn) [inline tests/core/data_layer/util.py:148]: is this test code? if so, how come it's not under `tests` and run by pytest?

**Comment by arvidn** [comment]
I'm quite ignorant of the design of the database schema, but it seems like it's much more complicated than it would need to be, and it looks like we do a *lot* of database queries, that ought to be just a single query.

I'm also suspicious that the tree hashes are not computed correctly. Is there a test that demonstrates the tree hashes?

File: chia/data_layer/data_layer_client.py
Comment (fchirica) [reply chia/data_layer/data_layer_client.py:26]: it's the field from here:
https://github.com/Chia-Network/chia-blockchain/blob/atari/chia/data_layer/data_store.py#L72
I suspect it's to differentiate between more tables stored in the same datastore.

File: chia/data_layer/data_layer_client.py
Comment (fchirica) [reply chia/data_layer/data_layer_client.py:31]: no, this is how it's stored in the DB:
https://github.com/Chia-Network/chia-blockchain/blob/atari/chia/data_layer/data_store.py#L72

File: chia/data_layer/data_layer_client.py
Comment (fchirica) [reply chia/data_layer/data_layer_client.py:85]: I have used the method from here:
https://github.com/Chia-Network/chia-blockchain/blob/atari/chia/data_layer/data_store.py#L143
but perhaps factor it out and call the same function in both places?

File: chia/data_layer/data_layer_server.py
Comment (fchirica) [reply chia/data_layer/data_layer_server.py:16]: it increments by 1 every time we insert or delete a single node - as far as I understand the code.

File: chia/data_layer/data_layer_server.py
Comment (fchirica) [reply chia/data_layer/data_layer_server.py:39]: no, the production code should read the actual DataStore DB that other services write to. however I needed to start somewhere, and creating my own DataStore demonstrate the client/server functionality. will think of creating a test out of this.

File: chia/data_layer/data_store.py
Comment (fchirica) [reply chia/data_layer/data_store.py:404]: this table is not immutable. suppose we get an update for x -> y, z to x' -> y', z. then z's ancestor changed from x to x'.

File: chia/data_layer/data_store.py
Comment (fchirica) [reply chia/data_layer/data_store.py:738]: this function returns left-to-right ordering of the tree. when we're the rightmost path in `node_hash` subtree, but still need to answer more nodes, we need to go up in the ancestors and then right, to find the next subtree.

File: chia/data_layer/data_store.py
Comment (arvidn) [reply chia/data_layer/data_store.py:404]: the update would hash to something other than the previous value, so both entries would exist after the update.
but maybe I'm misunderstanding this table, is this a table of *just* the peak table state?
(if so, I guess I don't understand why we need that, I would expect we could have an index on both key and generation/sequence number)

File: chia/data_layer/data_store.py
Comment (arvidn) [reply chia/data_layer/data_store.py:738]: so there is probably a better name than "answer_server_query()" then. So, do we return the subtree rooted in `node_hash` but then uncles from `node_hash` to the `root`, to prove `node_hash` is a part of the tree?
The term "ancestors" I thought meant previous states of the tree, is that what it refers to? or does it refer to parent and uncle nodes in the tree?

File: chia/data_layer/data_layer_client.py
Comment (arvidn) [reply chia/data_layer/data_layer_client.py:31]: then I would propose you simplify this (and make the intention clear):

```
node = root_hash = root_json["node_hash"]
```

File: chia/data_layer/data_store.py
Comment (fchirica) [reply chia/data_layer/data_store.py:404]: this is just a map from a node to its parent, so we can quickly get the path to the root.

File: chia/data_layer/data_store.py
Comment (fchirica) [reply chia/data_layer/data_store.py:738]: yes, it's parents and uncles. I'll rename it `get_left_to_right` ordering. no, the intend is not to prove node_hash is part of root. it's to get more nodes after we've finished the subtree. this saves queries: say we queried a subtree with 3 nodes, but if server wants 5000 answers, it'll go up and right in the tree to find other 4997 nodes.

File: chia/data_layer/data_store.py
Comment (fchirica) [reply chia/data_layer/data_store.py:534]: there can be multiple identical `keys` in the table - for different `tree_id`. so just finding the key isn't enough - we need to also check `tree_id`. anyway, I think this falls out of scope of this PR - I needed to make changes to insert only so I can test for a big tree. I'll change `optimized` to `skip_expansive_checks`.

File: chia/data_layer/data_store.py
Comment (altendky) [inline chia/data_layer/data_store.py:418]: I would think a recursive query like already in use would be much faster than bouncing back and forth between Python and SQLite?  I can certainly explain those queries if you want.  It took me awhile to get my head around them but after that they are ok to read..

File: chia/data_layer/data_store.py
Comment (altendky) [inline chia/data_layer/data_store.py:761]: Please no returning tuples.  That just leaves us hoping we're really careful to have the return and call in the same order and also leads to nested hints that we can't keep track of.  Can the `bool` not be handled by an exception?  If we really need two things we can write a small class using dataclasses.

File: chia/data_layer/data_layer_server.py
Comment (altendky) [inline chia/data_layer/data_layer_server.py:10]: I've wondered if we could use an existing 'server' mechanism instead of our own, but should we continue to use our own system for this as we do for other RPCs/APIs?

File: chia/data_layer/data_store.py
Comment (altendky) [inline chia/data_layer/data_store.py:787]: This looks like an RPC-ready serialized form that the data store should not be aware of.

File: chia/data_layer/data_store.py
Comment (altendky) [inline chia/data_layer/data_store.py:753]: What is this doing?

File: chia/data_layer/data_store.py
Comment (fchirica) [reply chia/data_layer/data_store.py:90]: I need an efficient `get_ancestors` for server to answer queries quickly. I can remove all its references from `data_store`, but keep it for my server, until we figure out how to include versions to it?

File: chia/data_layer/data_store.py
Comment (fchirica) [reply chia/data_layer/data_store.py:318]: yes - this is redundant with the DB schema. I'll remove it.

File: chia/data_layer/data_store.py
Comment (fchirica) [reply chia/data_layer/data_store.py:789]: produces the left to right tree traversal starting from `node_hash`. this way client can ask server for nodes exactly from the place it downloaded last. also server isn't limited to `node_hash` subtree - if subtree is out of nodes, but query size still fits more nodes - we'll go up and right to find a new subtree.

File: chia/data_layer/data_store.py
Comment (altendky) [reply chia/data_layer/data_store.py:90]: Just for the scalability test though, right?

Comment (fchirica) [reply chia/data_layer/data_store.py:90]: no - server calls `get_left_to_right_ordering`, which calls `get_ancestors`.

File: chia/data_layer/data_store.py
Comment (fchirica) [reply chia/data_layer/data_store.py]: you're right - it's at least twice as fast as before. can you please double check the query is correct? it passes against `skip_expansive_checks` (i.e. against old `get_ancestors` method).

**Review by altendky — CHANGES_REQUESTED** [review]
So I ended up with a bunch of comments here but the majority of them are just tweaks.  There are a few big picture things I'll point out here though as well.

I guess the biggest thing is separating out the performance improvements/fixes for getting kv pairs and ancestors.  I kind of skipped those bits for this review but need to go back through them to make sure I internalize what you have coded.  I think we can reasonably separate these things as long as we keep the tests here to a small number of nodes for now?

Another topic that's less about this PR as much as general design for the data store is the matter of needing to handle a bulk generation insertion, presently coded in `DataLayerClient`, and the ability to create a new generation including multiple changes.  I think our expectation is to somehow track the individual steps even if the singleton is only updated after several changes, as opposed to calculating the diff.  This could be done by retaining the one-change-per-generation limit as it is now or by allowing multiple changes and recording the changes separately.  Given the existing implementations around the present design, perhaps we should just stick with that for now.

Thanks @fchirica for all the work on this.  I'm glad you were able to jump into this so quickly and get it working.  My apologies for taking so long to really look it over and understand what you have done.

File: chia/data_layer/data_layer_server.py
Comment (altendky) [reply chia/data_layer/data_layer_server.py:17]: I guess a caveat to this is that if the data propagation allows for getting arbitrary generations then we'll have a sparsely populated list of generations on some clients.

File: chia/data_layer/data_layer_server.py
Comment (fchirica) [reply chia/data_layer/data_layer_server.py:24]: it currently doesn't work without `.value`: `TypeError: Object of type Status is not JSON serializable`

File: chia/data_layer/data_store.py
Comment (fchirica) [reply chia/data_layer/data_store.py:785]: yes.

File: chia/data_layer/data_store.py
Comment (fchirica) [reply chia/data_layer/data_store.py:355]: yes - this code assumes every operation increments the generation (as opposed to a batch incrementing the generation)

File: chia/data_layer/data_layer_server.py
Comment (fchirica) [reply chia/data_layer/data_layer_server.py]: I've changed it to websockets and passing around JSONs as `aiohttp.WSMsgType.TEXT`.

**Comment by altendky** [comment]
Triggering CI after the latest merge to `atari`.

**Review by altendky — DISMISSED** [review]
Needs automated tests.  `:]`  Approving so we can merge and start working with all the pieces together.

**Pull request merged**
